pkg/controller/sync/version/manager.go

```diff
@@ -301,7 +304,9 @@ func (m *VersionManager) writeVersion(qualifiedName util.QualifiedName) util.Rec
 	creationNeeded := len(metaAccessor.GetResourceVersion()) == 0
 
 	if creationNeeded {
-		createdObj, err := m.adapter.Create(obj)
+		// Make a copy to avoid mutating the map outside of a lock.
+		createdObj := obj.DeepCopyObject()
```

Comment:
I understand the change on line 226 as `List()` is used inside a poll loop and it could effectively create a new `List Object` on each call to `adapter.List()` in earlier code, which could be a lot.
However I do not understand the change from earlier `adapter.Create()` and `adapter.Update()`. A copy of the object could be made inside those earlier methods, rather then exposing the client outside.
Is this done to keep `client` only at one place (as in `VersonManager` only now), because the fix was needed for `List()` anyways ?

Reply:
The version adapters previously internalized the client because they were calling entirely different methods internally.  For example, the adapter for `PropagatedVersion` was calling `PropagatedVersion.Get` method which would only accept the specific object type.  I don't think there's a good reason to internalize the client now that the only thing that differs about the client is the object type (the client just accepts `pkgruntime.Object`).  Given that there is only a single caller for both methods, removing the encapsulation actually requires less code (since only one implementation is required rather than 2) and exposes the copy requirement such that its purpose can be documented in relation to the map storage.

Reply:
Thanks for the explanation.